### PR TITLE
Create TranslationPage for translating story content

### PR DESCRIPTION
**Pull request by e-wai**
## Notion ticket link

[Load all story contents into cells/table](https://www.notion.so/uwblueprintexecs/Load-all-story-contents-into-cells-table-9d7c0ffb11954bb4a05c85b18920d717)

## Implementation description
(Commits should make it easier to review code and pieces and is an ok description of implementation:)
* Add components for page: image icons, `Cell` (story content), `EditableCell` (story translation content)
* Create `TranslationPage` to house these components. About how to store the story content / translation content -> array was the easier choice as it's ordered, maps easily, and passing content id / line Index around requests meant you could set quickly using indices (without looping to find). Back half of the `StoryLine` object is optional as translated content set in a different loop from story content.
* There's a couple bugs around story translation content queries/mutations, this fixes one of them (that it wasn't saving the changes into db when called)

## Steps to test
1. From the stories page, go to a translations route -> pressing `view translation / edit translation / translate book` should lead to a `/translation/storyId/storyTranslationId` route ... some of these pages are broken
2. Edit text boxes to make the translation. You can only do this slowly - the changes don't register properly when you do it quickly. However, you can change it not one letter at a time by copy/pasting longer messages. Note that the icons flicker (they are appear randomly, without any logic yet).
3. You can check that the changes are being saved properly: the easiest way is to refresh the page and ensure that the changes made still appear (and are thus saved somewhere). You can also open up the db and check that changes are being registered properly.
translation load content

`<TextArea>` doesn't grow automatically, but reportedly `<Input>` doesn't do multi-line 

## What should reviewers focus on?
* I stayed fairly light on the CSS. Do we have finalized designs or is this mostly placeholder until the stylized components to finish up?
* More CSS: I used flex for rows instead of grid... is that a bad decision from a best practices / efficiency standpoint?
* Is there a preferred name for the translation page route?
* (I think we talked about this, but just noting -> this re-renders the entire page for every change and makes a new query every single time. That seems less than ideal)
* How do the dimensions look on your browser? 
* Should create a ticket to investigate and double check which queries/mutations are broken? I'm seeing some null errors with `StoryTranslationsAvailableForReview` and `StoryTranslationsByUser` 
* Are there error states to cover, and do we want a loading card?

## Checklist
- [x] My PR name is descriptive and in imperative tense
- [x] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [x] For backend changes, I have run the appropriate linters:  `docker exec -it <backend-container-id> /bin/bash -c "black . && isort --profile black ."`
- [x] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [inline frontend/src/components/pages/TranslationPage.tsx:155]: Proper order on the page assumes that `storyContent` will be ordered by lineIndex, which I'm not sure if I can guarantee unless that SQL query has a nice ORDER BY default. Should the array (since it's mapped) be sorted in the frontend or place an ORDER BY on the query endpoint?

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [inline frontend/src/components/pages/TranslationPage.tsx:164]: I kept getting `""` back from the backend, but this was registering as undefined here?

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [reply frontend/src/components/pages/TranslationPage.tsx:155]: Also, just double-checking - would altering the state object `translatedStoryLines` instead of using a local const `contentArray` force a re-render each iteration of `forEach`? I assumed it would but wasn't sure

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [inline frontend/src/components/pages/TranslationPage.tsx:178]: My big (?) here is whether I can avoid looping so much -> right now, I'm looping
- once over the storyContents response
- once over the storyTranslationContents response
- once over the combined object state array (which has everything from the above)

That's not so painful for my 10-element array, but could get gross for larger stories?

**Review by gaoxk — CHANGES_REQUESTED** [review]
Great work! really exciting to see this all come together. It's quite a challenging PR. 
### Re: What should reviewers focus on?

> I stayed fairly light on the CSS. Do we have finalized designs or is this mostly placeholder until the stylized components to finish up?

I think everything is still in wireframes. Can you run this pr by Shaahana and get her thoughts?
> More CSS: I used flex for rows instead of grid... is that a bad decision from a best practices / efficiency standpoint?

based on this article, I think grid may be a better choice https://webdesign.tutsplus.com/articles/flexbox-vs-css-grid-which-should-you-use--cms-30184
flex is for when you want to restack/adjust components easily if things are oversized or too long, but for the translation platform, I dont believe we ever want editing cells to sit on top of eachother if a cell is too long.  
CSS grids are more rigid/precise about where components go. 

However, this relies on designs, and I think we're better of leaving this decision for after usability testing. I'm happy with sticking with flex for now. 
> Is there a preferred name for the translation page route?

I think what you have is good!
> (I think we talked about this, but just noting -> this re-renders the entire page for every change and makes a new query every single time. That seems less than ideal)

hm. constant rerendering and requests per keystroke is worse in practice than I imagined. This article has a neat solution with hooks that I don't think requires too too much refactoring on the current pr https://www.synthace.com/autosave-with-react-hooks/  Open to discussing it more - if you think it's a significant change then we can split it off to a seperate ticket
> How do the dimensions look on your browser?

good, though there appears to be a chin on the box of cells/editables. 
image

> Should create a ticket to investigate and double check which queries/mutations are broken? I'm seeing some null errors with StoryTranslationsAvailableForReview and StoryTranslationsByUser

Are these the null translationcontent errors? or full stories coming back null?
> Are there error states to cover, and do we want a loading card?

hm. we can display an error modal if any saves/loads dont work but I'd want Shaahana's input first. Let's save error handling for another time.

File: frontend/src/components/pages/TranslationPage.tsx
Comment (gaoxk) [reply frontend/src/components/pages/TranslationPage.tsx:87]: ps: I'm open to indexing from either 0 or 1 - we'd just have to make sure appropriate changes to db script are made and the UI still indexes for 1 (i dont imagine most people are familiar with 0-indexing)

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [reply frontend/src/components/pages/TranslationPage.tsx:141]: Not sure how zipping would look. Without zipping, I'm thinking the options would be
- 2 sorts (both response arrays) + one iteration (go through both together by index)
- 1 sort (first response array) + 2 iterations (go through each separately)

I think second should be more efficient?

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [reply frontend/src/components/pages/TranslationPage.tsx:157]: Not sure how zipping would look. Without zipping, I'm thinking the options would be
- 2 sorts (both response arrays) + one iteration (go through both together by index)
- 1 sort (first response array) + 2 iterations (go through each separately)

I think second should be more efficient?

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [reply frontend/src/components/pages/TranslationPage.tsx:87]: Potentially resolved by https://github.com/uwblueprint/planet-read/pull/46

- Going with 0 indexing. Makes most sense from dev standpoint, and frontend implementation is just + 1

File: frontend/src/components/pages/TranslationPage.tsx
Comment (e-wai) [reply frontend/src/components/pages/TranslationPage.tsx:141]: ^ response on below comment: https://github.com/uwblueprint/planet-read/pull/39/files#r667644364

**Comment by e-wai** [comment]
> ### Re: What should reviewers focus on?
> > I stayed fairly light on the CSS. Do we have finalized designs or is this mostly placeholder until the stylized components to finish up?
> 
> I think everything is still in wireframes. Can you run this pr by Shaahana and get her thoughts?

Good for now. Designs given as 60-70% done
> 
> > (I think we talked about this, but just noting -> this re-renders the entire page for every change and makes a new query every single time. That seems less than ideal)
> 
> hm. constant rerendering and requests per keystroke is worse in practice than I imagined. This article has a neat solution with hooks that I don't think requires too too much refactoring on the current pr https://www.synthace.com/autosave-with-react-hooks/ Open to discussing it more - if you think it's a significant change then we can split it off to a seperate ticket

I'll push up clean up for this PR for now and work on this for the next step on top of this PR. If there's still a lot of outstanding changes, it might make it into this PR but otherwise I'm thinking it's worth its own PR.

> > How do the dimensions look on your browser?
> 
> good, though there appears to be a chin on the box of cells/editables.
> image

This was intentional, I think that's what the designs look like?

> > Should create a ticket to investigate and double check which queries/mutations are broken? I'm seeing some null errors with StoryTranslationsAvailableForReview and StoryTranslationsByUser
> 
> Are these the null translationcontent errors? or full stories coming back null?

just translation_content afaik

> > Are there error states to cover, and do we want a loading card?
> 
> hm. we can display an error modal if any saves/loads dont work but I'd want Shaahana's input first. Let's save error handling for another time.

Note from Slack: no real conclusion yet

File: frontend/src/components/translation/EditableCell.tsx
Comment (gaoxk) [reply frontend/src/components/translation/EditableCell.tsx:9]: Lets leave things as they are since you're going to be fixing up the react hook next.

File: frontend/src/components/pages/TranslationPage.tsx
Comment (gaoxk) [reply frontend/src/components/pages/TranslationPage.tsx:159]: I agree, second is more efficient

File: frontend/src/App.tsx
Comment (gaoxk) [inline frontend/src/App.tsx:40]: heads up: may need rebase if jenn merges first

**Review by gaoxk — COMMENTED** [review]
Re:Re

> just translation_content afaik

👍 Armanya is currently working on a fix

regarding the chin: I'll pass it by design just in case - I thought they were going for an infinite scroll

**Comment by e-wai** [comment]
Post-rebase 👀 

image

**Comment by gaoxk** [comment]
image
Looks like we don't want the chin, your call for whether or not you wanna put this in a separate pr

File: frontend/src/components/pages/TranslationPage.tsx
Comment (gaoxk) [inline frontend/src/components/pages/TranslationPage.tsx:15]: In the current state, the translation progress bar is static at 25%. Do you want to integrate the state or save this for a separate ticket?

Comment (jennchenn) [reply frontend/src/components/pages/TranslationPage.tsx:15]: If it makes it easier, I can work on implementing the rest of the progress bar logic once this PR is merged; will need to figure out how to update the progress on updates as well in the frontend

**Comment by e-wai** [comment]
One more CSS update -> no chin

translation load content final

**Review by gaoxk — APPROVED** [review]
Noice! Great work 🚢 

>Also pretty important note I realize that I didn't include. No top bar right now (which might be sticky? idk)

I saw the slack convo you had with Shaahana and I'll make a ticket for this - the top bar has many more features that need to come together that are out of scope for this pr

**Pull request merged**
